# <font color='#eb3483'> GETTING STARTED </font>


Welcome to the <font color='#eb3483'> iXperience course on Data Science</font>! We can't wait to meet you and kick off the class! In the meantime, please run through this notebook to familiarise yourself with Jupyter notebooks and also some Python coding basics.

### Some navigating in Jupyter Notebooks

In [1]:
# to run a cell hit Control+Enter
# to run a cell and move to the next cell hit Shift+Enter

print('Hello World!')


Hello World!


### Markdown

This is a markdown cell. Double click to see the code.

To change the cell type (and other options) you need to be in "command mode". You are in command mode when the colour on the left of your cell is blue and you are in "edit mode" when the colour is green.

To get into command mode you can hit esc, click the pencil icon on the top right or click next to the line colour on the left of the cell.

Try it in the empty cell below.

To get back to edit mode from command mode hit enter.

To change a cell to Markdown. Enter command mode and select markdown from the dropdown menu above or simply hit m.

To view command mode shortcuts - hit h while in command mode



Once a cell is set to output markdown, you can write text and it will turn into beautiful html. Here is a great markdown [cheatsheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code) to keep handy!

**This in bold** 
(double click this cell to see how we generate these effects - to transform to markdown simply run cell as normal

*This in italics*

### Introduction to Jupyter Header

- Element 1
  - Element 1.1
  
  
<hr>

![kitty](https://c24e867c169a525707e0-bfbd62e61283d807ee2359a795242ecb.ssl.cf3.rackcdn.com/imagenes/gato/etapas-clave-de-su-vida/gatitos/nuevo-gatito-en-casa/gatito-tumbado-lamiendo-sus-patitas.jpg)

<hr>

### Code cells

In [2]:
# This is a comment, python ignores these lines. Its useful to add small explanations to your code or notes to yourself 

#Python scripts are executed linearly, line after line, top to bottom

print('Hello World!')

print("Second line to print!")

# in python, text (strings), can be defined either with single quotes ' or double quotes "
print('Printing our third line')


Hello World!
Second line to print!
Printing our third line


### A few more helpful tips

When in command mode here are a few handy shortcuts. Hitting:
- m changes cell to markdown
- a inserts cell above
- b inserts cell below
- dd deletes a cell.
- z undoes delete cell


<hr>


# <font color='#eb3483'> INTRO TO PYTHON </font>

<font color='#eb3483'> 1. Variables in Python </font>

<font color='#eb3483'> 2. Data types </font>

<font color='#eb3483'> 3. Data Structures </font>

<font color='#eb3483'> 4. Control Flow types </font>

<font color='#eb3483'> 5. Functions </font>



<br>

**Be an active rather than a passive learner:**

Anybody can hit shift enter and whiz through this notebook. 

As you go through the notebook. Stop. Think. Can you think of a situation you might use that code.

Change inputs, change names ... see what happens. Be playful and inquisitive. You wont break anything. You can always download it again :)

If something doesnt make sense. Google for an explanation. 

If you arent sure what a piece of code does - try changing it and see. 

Or you can always use the inbuilt help function as in the cell below


In [3]:
print?

Signature: print(*args, sep=' ', end='\n', file=None, flush=False)
Docstring:
Prints the values to a stream, or to sys.stdout by default.

sep
  string inserted between values, default a space.
end
  string appended after the last value, default a newline.
file
  a file-like object (stream); defaults to the current sys.stdout.
flush
  whether to forcibly flush the stream.
Type:      builtin_function_or_method

**I recommend doing this notebook in a couple of sessions over a day or two to just to let the information sink in.**

This first notebook has a lot of the really basic need to knows.
The use of code in the examples below is to show you how each function works. It might seem a little bit abstract here. But once we nail down the basics we can start applying them to real data - and their utility will come shining through.

So for now - let's smash the basic and practice some python!


<br>
<br>

<hr>



# <font color='#eb3483'> 1. VARIABLES IN PYTHON </font>
In Python, variable names are like stickers put on objects. Every sticker has a unique name written on it, and it can only be on one object at a time. If desired, more than one sticker can be put on the same object.

In [4]:
# assign the 'animal' sticker to the cat string
animal = 'cat'

# also assign the 'pet' sticker to the underlying cat string
pet = animal

print(f'Animal is {animal}, pet is {pet}') # we'll get to the f annotation in a few seconds :) 
#But to see what the f does, delete it and try run the code again

# move the animal sticker to giraffe
animal = 'giraffe'

print(f'Animal is now {animal}, pet is still {pet}')

Animal is cat, pet is cat
Animal is now giraffe, pet is still cat


In [5]:
# How would you get pet to change as well? Try it out below!




# END OF VARIABLE TYPES 
(They arent all this quick ;) )

![variables](https://starecat.com/content/wp-content/uploads/actual-programming-vs-debating-30-minutes-on-how-to-name-a-variable-drake-meme.jpg)

<hr>
<br>
<br>
<br>


# <font color='#eb3483'> 2. DATA TYPES </font>
Python has many types of data that are a part of the core language, they 
are called primitives.

The most common primitives are 
1. strings, 
1. numbers (integers and decimals), and 
1. booleans

### <font color='#eb3483'> 1. Strings </font>

In [6]:
print("We use strings to represent text:")

var_str = 'Hello'
var_str2 = "World!"
print(var_str)
print(var_str2)
print(var_str, var_str2)

print(type(var_str2))

We use strings to represent text:
Hello
World!
Hello World!
<class 'str'>


In [7]:
#If you want to better understand what a function does use the inbuilt help feature
#by either typing ?function_name or using the help menu item (or good old google)

?print

Signature: print(*args, sep=' ', end='\n', file=None, flush=False)
Docstring:
Prints the values to a stream, or to sys.stdout by default.

sep
  string inserted between values, default a space.
end
  string appended after the last value, default a newline.
file
  a file-like object (stream); defaults to the current sys.stdout.
flush
  whether to forcibly flush the stream.
Type:      builtin_function_or_method

Strings can be joined to create longer strings. There are many ways to join strings
(this is called string interpolation).

We can use the simbol `+` to "sum" (concatenate) strings

In [8]:
name = "Elvis"
surname = "Presley" #(we call a last name a surname in South Africa)

var_str_joined = name + " " + surname #remove the space between the "" and see what happens
print(var_str_joined)

Elvis Presley


In [9]:
#We can also "multiply" strings
n = "hello"
print(n * 10)

hellohellohellohellohellohellohellohellohellohello


We can also use the method `format()` that allows us to insert strings inside other strings


In [10]:
name = "Spock"
city = "Shi'Kahr"

presentation = "Hi, I'm {}, and I'm from {}"

print(presentation.format(name, city))


Hi, I'm Spock, and I'm from Shi'Kahr


Try make variables of your own name and city in the below cell
and use the presentation command above to print your name and city in a sentance.

In [11]:
name = "molly"
city = "charlotte"

presentation = "Hi im {} and im from {}"

print(presentation.format(name,city))

Hi im molly and im from charlotte


Python is a language with "batteries included" this means it includes a lot of 
functionalities when we install python.

Those functionalities are usually grouped in packages (or modules) and we can 
"load" them and use them with the keyword `import`

For example, if we want to use the `math` module, that has many mathematical functions,
we can access it if we type `import math`

In [12]:
import math

In [13]:
# Now we can access all the functionalities in the math module
print(math.pi)

3.141592653589793


When using the `format` method we can specify how do we want to format the 
variables. For example, we can round the variables.

In [14]:
pi_str = "The first digits of pi are: {}".format(math.pi)
print(pi_str)

The first digits of pi are: 3.141592653589793


In [15]:
# We can print the string rounded to 2 digits
pi_str = "The first digits of pi are: {:.3}".format(math.pi)
print(pi_str)

The first digits of pi are: 3.14


In [16]:
#Another (shorter) method to do string interpolation is referencing the variable 
#directly in the string by using an f-string

presentation2 = f"Hi, I'm {name}, and I'm from {city}"

print(presentation2)

Hi, I'm molly, and I'm from charlotte


In [17]:
#Checking we understand object naming

#Why does the below code give us that output
name = "John"
print(presentation2)

Hi, I'm molly, and I'm from charlotte


### More operations with strings

In [18]:
title = "introduction to PYTHON"

print("We can convert a string to uppercase with upper()")
print(title.upper())

print("We can convert a string to lowercase with lower()")
print(title.lower())

print("We can convert the first letter to upper case with capitalize()")
print(title.capitalize())

We can convert a string to uppercase with upper()
INTRODUCTION TO PYTHON
We can convert a string to lowercase with lower()
introduction to python
We can convert the first letter to upper case with capitalize()
Introduction to python


In [19]:
name_with_commas = ",Connor,"

print("We can use `strip()` to remove characters from the beginning and the end of a string")
print(name_with_commas.strip(","))

We can use `strip()` to remove characters from the beginning and the end of a string
Connor


In [20]:
print("We use `replace()` to replace parts of a string for something else")
print(name_with_commas.replace("nor", "man"))

# think of an example where you might need to use replace

We use `replace()` to replace parts of a string for something else
,Conman,


In [21]:
print("We can chain methods together")
print(
      name_with_commas
      .strip(",")
      .replace("nor", "man")
      .upper())


We can chain methods together
CONMAN


In [22]:
# We can split a string in multiple strings by using `split()`

email = "spongebob.squarepants@ixperience.co"
email_bits = email.split("@")
print(email_bits)

# think of a situation where you might need to use split.

['spongebob.squarepants', 'ixperience.co']


### <font color='#eb3483'> Quick knowledge check! </font>
1. Your friend just sent you his part of an essay (see "essay" variable below) but you realized it's a mess! Write some code to correct the contraction (i.e. doesn't -> does not), properly capitalize the sentence, and take out all of the times he says totally!

In [23]:
essay = "sometimes the world can be totally messed up but it totally doesn't have to be!"

print(
    essay.replace(" totally ", " ").replace("doesn't", "does not").capitalize())


Sometimes the world can be messed up but it does not have to be!


2. You just got the results back from your student body election (obviously analyzed through Python), print out a nice final statement summarizing the results (whatever style floats your boat - play around and have fun with it!)

In [24]:
winner = "Pedro"
loser = "Summer"
pedro_vote_percentage = 60.3594594594

print(f"the winner is {winner} and the loser is {loser}, by {pedro_vote_percentage} percent!")

the winner is Pedro and the loser is Summer, by 60.3594594594 percent!


<hr>
<br>

### <font color='#eb3483'> 2. Numbers </font>

In [25]:
print("There are two basic numeric primitives in python, int(integers) and float(decimals)")

integer = 23
print(type(integer))

#%%
decimal = 23.1
print(type(decimal))

#%%
print(type(23.))

There are two basic numeric primitives in python, int(integers) and float(decimals)
<class 'int'>
<class 'float'>
<class 'float'>


In [26]:
print("We can convert numbers to strings easily")
print(type(str(decimal)))
print("two + " + str(decimal))

We can convert numbers to strings easily
<class 'str'>
two + 23.1


In [27]:
print("We can also use numbers in string interpolation")
name = "Taryn"
city = "Cape Town"
count = 145000
print(f"Hi, I'm {name}, I'm from {city} and I've been in lockdown for {count} days")

We can also use numbers in string interpolation
Hi, I'm Taryn, I'm from Cape Town and I've been in lockdown for 145000 days


In [28]:
print("Similarly, we can convert strings to numbers")
number_string = "24" # remember it is a string because it is captured with ""
print(int(number_string) + 5)
print(float(number_string) + 5)

Similarly, we can convert strings to numbers
29
29.0


In [29]:
print("We have to make sure the numbers are valid!")
invalid_number_string = "24.5"
print(float(invalid_number_string))

We have to make sure the numbers are valid!
24.5


In [30]:
#Try changing the above so that it is valid

### Operations with numbers

We can use the basic arithmetic symbols to perform operations with numbers


In [31]:
#sum
print("2+2=",2+2)

#substraction
print("4-9=",4-9)

#multiplication
print("3*2=",3*2)

#division
print("7/2=",7/2)

2+2= 4
4-9= -5
3*2= 6
7/2= 3.5


In [32]:
#There are other useful operations too

a = 7
b = 2
# floor division, do a division and remove the remainder --> //
print(f"{a}//{b}=", a//b)

# modulo, do a division and return the remainder --> %
print(f"{a}%{b}=", a%b)

#negation, change the sign of a number
print(f"negation of {a}=", -a)

#square, cube, fourth power, etcetera
print(f"{a}**{b}=",a**b)

7//2= 3
7%2= 1
negation of 7= -7
7**2= 49


### <font color='#eb3483'> Quick knowledge check! </font>
1. Whoops, we can't seem to figure out why this code isn't working. Please fix it for us! (Bonus: Can you do it without changing how we defined x and y?)

In [33]:
x = "24.5"
y = 13
mySum= float(x) + y
print('X + Y:',str(mySum))

X + Y: 37.5


<hr>
<br>

### <font color='#eb3483'> 3. Booleans </font>
A boolean variable is a primitive that can only be True or False


In [34]:
truth = True
falsehood = False

print(type(truth))

<class 'bool'>


As an additional primitive we have `None` which is the null variable


In [35]:
null = None
print(type(null))

<class 'NoneType'>


In [36]:
#Most variables can be converted to a boolean, usually they are converted to True
print(bool("potato"))

True


In [37]:
# The null values (0, "", None, and empty lists and dicts) are converted to False

print(bool(""))
print(bool(0))
print(bool(null))
print(bool([]))
print(bool({}))

False
False
False
False
False


### Logic Comparisons

We can compare variables, the result of a comparison is always a boolean

In [38]:
a = 7
b = 2
# a greater or equal to b
print(f"{a} > {b}", a > b)
# b smaller or equal to a
print(f"{b}<={a}", b <= a)
# b equals 2
print(f"{b}==2", b == 2)
# a is different than 23
print(f"{a}!=23", a != 23)

7 > 2 True
2<=7 True
2==2 True
7!=23 True


We can evaluate the opposite of a logic comparisons with `not`

In [39]:
print(f"not {a}!=23 --> " , not a != 23)

not 7!=23 -->  False


We can evaluate that multiple conditions are true with `and`

In [40]:
print(f"{a}!=23 and {a}<{b} --> ", a != 23 and a < b)

7!=23 and 7<2 -->  False


We can evaluate that any condition is true with `or`


In [41]:
print(f"{a}!=23 or {b}<{a} --> ", a != 23 or a < b)

7!=23 or 2<7 -->  True


When we want to check if something is True or False, we dont use `==`, we use `is`

In [42]:
print("truth is True=", truth is True)


truth is True= True


### <font color='#eb3483'> Quick knowledge check! </font>
1. What should the following code evaluate to? Try figuring it out by hand and then run the code to find out!

In [43]:
statement = not(((7 > 3) and (bool(""))) or (('a'!='b') or (1 <= 0)))

print('The statement is', statement)

The statement is False


# BOOM THE END OF DATA TYPES

![dt](https://culshoefer.gitbooks.io/programming-is-fun-with-python-and-pygame/content/img/boolean-to-integer.jpg)



<hr>
<br>
<br>
<br>
<br>

# <font color='#eb3483'> 3. DATA STRUCTURES </font>

Data structures are those objects we use to store information (data)
There are many different types, here we are going to see the basics of
1. Lists 
2. Tuples
3. Dictionaries


### <font color='#eb3483'> 1. Lists </font>

Lists of one of the fundamental data structures in Python, and can be described generally as an ordered collection (list) of objects. At their most basic level, lists have a length (len(list)), and each element in the list can be accessed via its unique integer index.

Note that in Python, all data structures are zero-indexed, meaning that the first occuring entry is assigned an index of 0 instead of 1.

In [44]:
fruits = ["orange", "apple", "pear", "strawberry", "pineapple"]

print(fruits)

print(type(fruits))

['orange', 'apple', 'pear', 'strawberry', 'pineapple']
<class 'list'>


We access elements in a list with brackets []
 The index (position) for python lists starts at 0, this means we access the first
 element by [0]

In [45]:
first_fruit = fruits[0]
print(first_fruit)

orange


In [46]:
# We can access a range of elements by doing [start:end-1:order]
print(fruits[:2])

['orange', 'apple']


In [47]:
# get the elements after the second one
print(fruits[2:])

['pear', 'strawberry', 'pineapple']


In [48]:
"""We can select ranges starting  from the last one by using negative numbers
For example, we can get the last 2 elements by doing:
"""
print(fruits[-2:])

['strawberry', 'pineapple']


In [49]:
# We can skip elements by using [::n], for example, if we just want to select the odd elements:
print(fruits[::2])

['orange', 'pear', 'pineapple']


In [50]:
# if we use a negative number  for the order we will get the list reversed
print(fruits[::-1])

['pineapple', 'strawberry', 'pear', 'apple', 'orange']


In [51]:
# We can see the number of elements in a list with `len`
print(len(fruits))

5


In [52]:
# We can add elements to the end of a list with `append()`.
fruits.append("melon")
print(fruits)

['orange', 'apple', 'pear', 'strawberry', 'pineapple', 'melon']


In [53]:
fruits.append(5)
fruits.extend(["banana", "grape"])
print(fruits)

['orange', 'apple', 'pear', 'strawberry', 'pineapple', 'melon', 5, 'banana', 'grape']


In [54]:
# we can repeat a list multiplying it by a number
print(fruits * 2)

['orange', 'apple', 'pear', 'strawberry', 'pineapple', 'melon', 5, 'banana', 'grape', 'orange', 'apple', 'pear', 'strawberry', 'pineapple', 'melon', 5, 'banana', 'grape']


In [55]:
# We can add lists together with `+`
cities = ["Cape Town", "New York"]
print(fruits + cities) # Prints concatenated lists

['orange', 'apple', 'pear', 'strawberry', 'pineapple', 'melon', 5, 'banana', 'grape', 'Cape Town', 'New York']


In [56]:
# We can modify elements in a list

fruits[0] = "mango"
print(fruits)

['mango', 'apple', 'pear', 'strawberry', 'pineapple', 'melon', 5, 'banana', 'grape']


In [57]:
# We can extend a list regardless of its initial size
fruits[3:] = ["litchi", "fig", "pommegranate", "watermelon"]
print(fruits)

['mango', 'apple', 'pear', 'litchi', 'fig', 'pommegranate', 'watermelon']


In [58]:
vegetables = ["tomato", "lettuce"]
# we can check if an element exists on a list
print(fruits)
print("tomato" in vegetables)

print("tomato" in fruits)

['mango', 'apple', 'pear', 'litchi', 'fig', 'pommegranate', 'watermelon']
True
False


In [59]:
# We can find the position of an element in a list

fruits = ["orange", "apple", "pear", "strawberry"]

print(fruits)

print(fruits.index("pear"))

print("Position in the list for the word pear is {}".format(fruits.index("pear")))

['orange', 'apple', 'pear', 'strawberry']
2
Position in the list for the word pear is 2


In [60]:
# we can remove an element at a specific position from a list with `pop()`

a = fruits.pop(2)
print(fruits)
print(a)

['orange', 'apple', 'strawberry']
pear


In [61]:
# So combining index and pop we can remove a specific element from a list
index_orange = fruits.index("orange")
fruits.pop(index_orange)
print(fruits)

['apple', 'strawberry']


In [62]:
# lists can be sorted with the method `sort()`

ages = [23, 33, 10,54,65,34,25]
ages.sort()
print(ages)

[10, 23, 25, 33, 34, 54, 65]


In [63]:
fruits = ["orange", "apple", "pear", "strawberry"]

print(fruits.sort())
print(fruits)

None
['apple', 'orange', 'pear', 'strawberry']


In [64]:
# We can generate lists of numbers with `range()`
range(10)
print(list(range(10)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [65]:
# strings can be considered lists of letters, and can be accessed same the way as lists

name = "Taryn"

print(name[0])
print(name[2:])

T
ryn


In [66]:
len(name)


5

In [67]:
#We can convert lists to strings with `.join()`

fruits = ["orange", "apple", "pear", "strawberry"]

comma_separated_fruits = ",".join(fruits)

print(comma_separated_fruits)
print(type(comma_separated_fruits))

orange,apple,pear,strawberry
<class 'str'>


In [68]:
line_separated_fruits = "\n".join(fruits)

print("Shopping list: \n{}".format(line_separated_fruits))

Shopping list: 
orange
apple
pear
strawberry


### <font color='#eb3483'> Quick knowledge check! </font>
1. Make a list with your top 5 favorite snacks

In [2]:
snacks = ["apples", "cookies", "toast", "pizza", "berries"]

2. Print out your list in alphabetical order, and reverse alphabetical order

In [11]:
snacks = ["apples", "cookies", "toast", "pizza", "berries"]
snacks.sort()
print(snacks)

sorted(snacks, reversed=True)
print(snacks)

['apples', 'berries', 'cookies', 'pizza', 'toast']


TypeError: 'reversed' is an invalid keyword argument for sort()

3. Write code to get the position of your least favorite element and remove it from the list

### <font color='#eb3483'> 2. Tuples </font>

Tuples are like lists, but once created can't be modified


In [69]:
musketeers = ("Athos", "Porthos", "Aramis")
print(type(musketeers))
print(musketeers)

<class 'tuple'>
('Athos', 'Porthos', 'Aramis')


In [70]:
# We can access tuple elements like we do with lists
print(musketeers[1:])

('Porthos', 'Aramis')


In [71]:
# however, we can't modify a tuple
musketeers[3] = "D'artagnan"

TypeError: 'tuple' object does not support item assignment

### <font color='#eb3483'> 3. Dictionaries </font>


Dictionaries are pairs of keys associated to values. We can easily (and quickly) see a value if we know its associated key

In [72]:
inventory = {
        "peaches": 3,
        "strawberries": 1,
        "apples": 4
        }
print(type(inventory))
print(inventory)

<class 'dict'>
{'peaches': 3, 'strawberries': 1, 'apples': 4}


In [73]:
# We can see all the keys in a dictionary with `keys()`
print(inventory.keys())

dict_keys(['peaches', 'strawberries', 'apples'])


In [74]:
# the method `keys()` doesnt return a list, if we want to access the keys like a 
# list we need to convert them to a list

# this will fail with a TypeError
print(inventory.keys()[0])

TypeError: 'dict_keys' object is not subscriptable

In [75]:
# this will work
print(list(inventory.keys())[0])

peaches


In [ ]:
# Similarly, we can see the values in a dictionary with `values()`
print(inventory.values())

In [ ]:
# We can read the value associated with a specific key with brackets
print(inventory["strawberries"])

In [ ]:
# If the key doesnt exist it will throw an error (KeyError)
print(inventory["melon"])

In [ ]:
# We can add a new key/modify an existing one to a dictionary using brackets

inventory["peaches"] = 1
inventory["oranges"] = 2
print(inventory)

In [ ]:
# We can check if a key exists in a dictionary
print("melon" in inventory)
print("peaches" in inventory)

In [ ]:
# We can remove a key from a dictionary with `pop()`

kilos_strawberries = inventory.pop("strawberries")
print("We have {} kilos of strawberries".format(kilos_strawberries))
print(inventory)

Each data structure can store other data structures!


In [ ]:
# a list with lists inside
flights =[
        ["Cape Town", "New York City"],
        ["Cape Town", "Johannesburg"],
        ["Cape Town", "Singapore"],
        ["New York City", "Paris"]
        ] 

print(flights[1][1])

In [ ]:
# a dictionary with tuples and lists as values
dict_flights = {
        "Cape Town": ("London", "Paris", "Berlin"),
        "Lisbon": ("New York City", "Madrid")
        }
print(dict_flights["Cape Town"])

In [ ]:
# a list containing dictionaries
dictionaries_list = [
        {"origin": "Cape Town", "destination": "New York"},
        {"origin": "Madrid", "destination": "Paris"}
        ]

print(dictionaries_list)

### <font color='#eb3483'> Quick knowledge check! </font>
1. Create a dictionary that stores phone contact information (i.e. name, phone number, company)

2. Write code that adds a key-value pair to store a <b>list</b> of the contact's social media handles

3. Nevermind, there's some privacy laws preventing us from storing social media handles. Write code to remove that key-value pair, make sure to print out the list of keys in the dictionary before and after to make sure you did your job!

# BOOM THE END OF DATA STRUCTURES

![lists](https://i.pinimg.com/originals/41/05/82/41058209535e7e800b7c06f1aee41df8.jpg)

<hr>

<br>
<br>
<br>


# <font color='#eb3483'> 4. CONTROL FLOW  </font>

Here we talk about control flow, which is the way that we can modify the linear
behaviour  (line by line) in python programs:

1. If-else
2. For loops
3. While loops

### <font color='#eb3483'> 1. If-else  </font>


We use if-else to take decisions and execute different parts of the code 
depending on a condition. 

In python, the way we declare we are inside a loop (or a function) is by 
indenting,
that is, if we are inside an if-else loop we put the code inside the loop 4 
spaces to the right. We can use the tab key in Jupyter Notebooks.

In [ ]:
# We can use an if loop to execute something based on one condition
temperature = 95

if temperature <= 65:
    #we indent this line to declare that it is inside the if loop
    print("Better get a sweater, its chilly out")
    
print("this gets printed no matter what")

In [ ]:
#If we want to add many different conditions to an if loop we can use if-else
temperature = 50


if temperature <= 30:
    mom_answer = "Get a coat, its freezing outside!"
    
elif temperature <= 65:           # elif stands for "else if ..."
    mom_answer = "Get a sweater, its chilly out!"
    
elif temperature <= 85:
    mom_answer = "Don't come back late!"
    
else:                             # The else block will run only if none of the previous conditions take place
    mom_answer = "Its too hot out!"

print(mom_answer)

In [ ]:
# try changing the temperature above and checking the result


In [ ]:

# We can insert loops inside other loops (nested loops)

temperature = 75
rain = False

if 65 < temperature and temperature < 85:
    # True evaluates to True, using not we get the opposite
    if not rain:
        print("Let's go picnic!")


### <font color='#eb3483'> Quick knowledge check! </font>
1. Write some code that lets you set your mood (i.e. happy) and recommends a song based on your mood.


### <font color='#eb3483'> 2. For Loops  </font>

FOR loops are used to iterate the elements of a list one by one(or anything that 
is iterable)

In [ ]:
numbers = [1, 2, 3, 4, 5, 6, 7, 80, 9, 10]

for n in numbers:
    if n <= 10:
        print(f"valid number {n}")
    else:
        print(f"ERROR! number {n} greater than 10")


In [ ]:
# Sometimes we want to break a for loop if a condition happens, we can break a loop with the keyword `break`

numbers = [1, 2, 3, 4, 5, 6, 7, 80, 9, 10]

for n in numbers:
    if n <= 10:
        print(f"valid number {n}")
    else:
        print(f"ERROR! number {n} greater than 10. EXITING THE LOOP")
        break

In [ ]:
# There are times where we dont want to break a loop, we just want to continue if
# something happens and do nothing, we can use `pass` for this

numbers = [1, 2, 3, 4, 5, 6, 7, 80, 9, 10]

for n in numbers:
    if n <= 10:
        print(f"valid number {n}")
    else:
        # pass is a placeholder, we have to put something inside an else block
        # we use pass so nothing happens.
        pass

In [ ]:
# continue, opposite to pass, will skip this loop iteration and go to the next one

numbers = [1, 2, 3, 4, 5, 6, 7, 80, 9, 10]

for n in numbers:
    if n <= 10:
        print(f"valid number {n}")
    else:
        continue
        print("this will never get printed")

In [ ]:
# A simplified way to iterate elements in a list is using a list comprehension

[number for number in numbers if number <= 10]

In [ ]:
#Make a line that writes "too big" for numbers >10
[number for number in numbers if number >= 10]

In [ ]:
# We can iterate the keys in a dictionary with a for loop

inventory = {
        "peaches": 3,
        "strawberries": 1,
        "apples": 4
        }

for fruit in inventory:
    print(fruit)

In [ ]:

# We can iterate the keys and values together if we use `.items()`:
for fruit, amount in inventory.items():
    print(fruit, amount)
    print("We have {} kilo/s of {}".format(amount, fruit))

In [ ]:
# We can iterate strings too! (remember, strings are just lists of characters).
name = "IXPERIENCE"

for letter in name:
    print("GIVE ME A {}!".format(letter))

### <font color='#eb3483'> Quick knowledge check! </font>
1. Using the animals dictionary below, write a for loop that loops through and prints out the scientific names of all the animals.

In [ ]:
animals = {
    'Elephant': {
        'scientific_name': 'Loxodonta Africana',
        'big_five': True
    },
    'Zebra':{
        'scientific_name': 'Equus Quagga',
        'big_five': False
    },
    'Black Rhinoceras':{
        'scientific_name': 'Diceros Bicornis',
        'big_five': True
    }
}

for animal in animals.keys():
    print(animals[animal]['scientific_name'])

2. Extension! Only print them out if they're big five animals.

### <font color='#eb3483'> 3. While loops  </font>
While loops continue running until a condition takes place



In [ ]:
n_beer_bottles = 99
while n_beer_bottles > 0:
    print("""
        🎜🎝♩ {} bottles of beer on the wall, {} bottles of beer.
        Take one down and pass it around, {} bottles of beer on the wall.🎜🎝♩
        """.format(n_beer_bottles, n_beer_bottles, n_beer_bottles - 1))
    n_beer_bottles -= 1

We need to be carefull when using a while loop, we might get stuck in an 
 INFINITE LOOP!
 (hit the stop button above or ii to exit the execution)

In [ ]:
n = 1
while n > 0:
    print("Stuck in the loop!")

A common use case for a while loop is to validate user input
 we can get user input with the function `input()`

In [ ]:
while True:
    input_user = input("Write a number from 1 to 10, type 'exit' to exit: ")
    continue
    try:
        if input_user == "exit":
            print("Bye!")
            break
        elif int(input_user) <= 10:
            square = int(input_user) ** 2
            print("The square of number {} is {}".format(input_user, square))
        else:
            print("the value {} is not valid".format(input_user))
    except ValueError:
        print("Error: the value {} can't be converted to an integer".format(input_user))

### <font color='#eb3483'> Quick knowledge check! </font>
1. Let's build a little app out of your music recommendation code from the if lesson block! Write a while loop that takes user input on their emotion and outputs a song recommendation until they say exit. (Challenge - can you make it recommend a different song if you pick the same emotion twice?)

# BOOM END OF CONTROL FLOW

![chips](https://data.whicdn.com/images/1093046/original.png)

<hr>
<br>
<br>
<br>


# <font color='#eb3483'> 5. FUNCTIONS  </font>

Functions are a way of packaging together code that does a specific job and that we might want to re-use. Let's say we wanted to write code that gets two radiuses and compares which has the bigger area.

In [ ]:
radius_1 = 10
radius_2 = 30

area_1 = 3.2 * radius_1**2
area_2 = 3.2 * radius_2**2

print('Area 1 is bigger than area 2: ', area_1 > area_2)

Instead of writing out $3.2*r^2$ twice, we could wrap it in a function!

In [ ]:
#To create a function  we use the keyword 'def' (as in define) and add the name of our function
#Next to the name of the function we add any input arguments (i.e. radius)
def computeArea(radius):
    #Return tells the function what to return (in this case the area)
    return 3.2*radius**2

For our example oru function is called computeArea and accepts one input argument, radius, and it returns the area. Let's see it in action!

In [ ]:
radius_1 = 10
radius_2 = 30

#Now instead of re-writing area calculation we can just use our function!
area_1 = computeArea(radius_1)
area_2 = computeArea(radius_2)

print('Area 1 is bigger than area 2: ', area_1 > area_2)

This might not seem like big savings, but what if had a more complicated function or wanted to use it in a bunch of different areas? It also let's us make changes all in one place - for example, I totally just remembered that Pi is 3.14 not 3.2!  Now we only have too make a change in one place (go ahead, do it yourself!)

In [ ]:
#Functions don't always need input arguments or return arguments
def sayHi():
    print('Hi')

sayHi()

In [ ]:
#Functions can also have multiple output arguments (just add them all in a comma separated list) too!
def capitalizeMyName(firstName, lastName):
    return firstName.capitalize(), lastName.capitalize()

#Notice that the results are returned in a tuple!
capitalizeMyName('warren', 'buffet')

Don't worry, python will remind you if you didn't input the right number of arguments!

In [ ]:
def myFun(foo, bar):
    return foo + bar

myFun()

Python is really great about handling all kinds of types, so your inputs and outputs to functions can be pretty much anything! Lists, dictionaries, even other functions!

In [ ]:
def metaFunctionBuilder(multiplier):
    #This function builds another function that is a multiplier
    def fun(x):
        return multiplier*x
    return fun

#We can assign the new function to a variable that let's us access it
newFunction = metaFunctionBuilder(5)

#Look it works!
newFunction(10)

We can also put default arguments for functions (this is great for when we have arguments that don't change most of the time)

In [ ]:
def multiply(x, multiplier = 10):
    return x*multiplier

print('With Defaults:', multiply(5)) 
print('With Multiplier = 5:', multiply(5, multiplier = 5)) #Note we can specify the parameter with the name
print('With Multiplier = 5:', multiply(5, 5)) #Or just put it in the right order

### <font color='#eb3483'> Quick knowledge check! </font>
1. Write a function that calculates the area of a rectangle. 

# BOOM END OF FUNCTIONS
![functions](https://i.pinimg.com/564x/7c/6a/ca/7c6acaabf60032c86abd6951c6cfc557.jpg)


# THAT'S IT! YOU SLAYED YOUR PREWORK STARTER NOTEBOOK. NOW HEAD ON OVER TO YOUR PREWORK ASSIGNMENT...

